File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/JoinKaigiInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/** Typed builder for {@code JoinKaigi} instructions. */
public final class JoinKaigiInstruction implements InstructionTemplate {

  private static final String ACTION = "JoinKaigi";

  private final KaigiInstructionUtils.CallId callId;
  private final String participant;
  private final String commitment;
  private final String commitmentAliasTag;
  private final String nullifierDigest;
  private final Long nullifierIssuedAtMs;
  private final String rosterRoot;
  private final String proofBase64;
  private final Map<String, String> arguments;

  private JoinKaigiInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private JoinKaigiInstruction(final Builder builder, final Map<String, String> argumentOrder) {
    this.callId = builder.callId;
    this.participant = builder.participant;
    this.commitment = builder.commitment;
    this.commitmentAliasTag = builder.commitmentAliasTag;
    this.nullifierDigest = builder.nullifierDigest;
    this.nullifierIssuedAtMs = builder.nullifierIssuedAtMs;
    this.rosterRoot = builder.rosterRoot;
    this.proofBase64 = builder.proofBase64;
    this.arguments = Map.copyOf(argumentOrder);
  }

  public KaigiInstructionUtils.CallId callId() {
    return callId;
  }

  public String participant() {
    return participant;
  }

  public String commitment() {
    return commitment;
  }

  public String commitmentAliasTag() {
    return commitmentAliasTag;
  }

  public String nullifierDigest() {
    return nullifierDigest;
  }

  public Long nullifierIssuedAtMs() {
    return nullifierIssuedAtMs;
  }

  public String rosterRoot() {
    return rosterRoot;
  }

  public String proofBase64() {
    return proofBase64;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static JoinKaigiInstruction fromArguments(final Map<String, String> arguments) {
    final Builder builder = builder();
    builder.setCallId(KaigiInstructionUtils.parseCallId(arguments, "call"));
    builder.setParticipant(KaigiInstructionUtils.require(arguments, "participant"));

    final String commitmentValue = arguments.get("commitment.commitment");
    if (commitmentValue != null) {
      builder.setCommitmentLiteral(commitmentValue);
      final String alias = arguments.get("commitment.alias_tag");
      if (alias != null) {
        builder.setCommitmentAliasTag(alias);
      }
    }

    final String nullifier = arguments.get("nullifier.digest");
    if (nullifier != null) {
      builder.setNullifierDigest(nullifier);
      final Long issuedAt =
          KaigiInstructionUtils.parseOptionalUnsignedLong(
              arguments.get("nullifier.issued_at_ms"), "nullifier.issued_at_ms");
      builder.setNullifierIssuedAtMs(issuedAt);
    }

    builder.setRosterRootLiteral(arguments.get("roster_root"));
    builder.setProofBase64(arguments.get("proof"));

    return new JoinKaigiInstruction(builder, new LinkedHashMap<>(arguments));
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof JoinKaigiInstruction other)) {
      return false;
    }
    return Objects.equals(callId.domainId(), other.callId.domainId())
        && Objects.equals(callId.callName(), other.callId.callName())
        && Objects.equals(participant, other.participant)
        && Objects.equals(commitment, other.commitment)
        && Objects.equals(commitmentAliasTag, other.commitmentAliasTag)
        && Objects.equals(nullifierDigest, other.nullifierDigest)
        && Objects.equals(nullifierIssuedAtMs, other.nullifierIssuedAtMs)
        && Objects.equals(rosterRoot, other.rosterRoot)
        && Objects.equals(proofBase64, other.proofBase64);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        callId.domainId(),
        callId.callName(),
        participant,
        commitment,
        commitmentAliasTag,
        nullifierDigest,
        nullifierIssuedAtMs,
        rosterRoot,
        proofBase64);
  }

  public static final class Builder {
    private KaigiInstructionUtils.CallId callId;
    private String participant;
    private String commitment;
    private String commitmentAliasTag;
    private String nullifierDigest;
    private Long nullifierIssuedAtMs;
    private String rosterRoot;
    private String proofBase64;

    private Builder() {}

    public Builder setCallId(final String domainId, final String callName) {
      this.callId = new KaigiInstructionUtils.CallId(domainId, callName);
      return this;
    }

    public Builder setCallId(final KaigiInstructionUtils.CallId callId) {
      this.callId = Objects.requireNonNull(callId, "callId");
      return this;
    }

    public Builder setParticipant(final String participant) {
      if (participant == null || participant.isBlank()) {
        throw new IllegalArgumentException("participant must not be blank");
      }
      this.participant = participant;
      return this;
    }

    public Builder setCommitment(final byte[] commitment) {
      this.commitment = KaigiInstructionUtils.canonicalizeOptionalHash(commitment);
      return this;
    }

    public Builder setCommitment(final String commitmentHexOrLiteral) {
      this.commitment = KaigiInstructionUtils.canonicalizeOptionalHash(commitmentHexOrLiteral);
      return this;
    }

    Builder setCommitmentLiteral(final String literal) {
      this.commitment = literal;
      return this;
    }

    public Builder setCommitmentAliasTag(final String aliasTag) {
      this.commitmentAliasTag = aliasTag;
      return this;
    }

    public Builder setNullifierDigest(final byte[] digest) {
      this.nullifierDigest = KaigiInstructionUtils.canonicalizeOptionalHash(digest);
      return this;
    }

    public Builder setNullifierDigest(final String digestHexOrLiteral) {
      this.nullifierDigest = KaigiInstructionUtils.canonicalizeOptionalHash(digestHexOrLiteral);
      return this;
    }

    public Builder setNullifierIssuedAtMs(final Long issuedAtMs) {
      if (issuedAtMs != null && issuedAtMs < 0) {
        throw new IllegalArgumentException("nullifier issuedAtMs must be non-negative");
      }
      this.nullifierIssuedAtMs = issuedAtMs;
      return this;
    }

    public Builder setRosterRoot(final byte[] rosterRoot) {
      this.rosterRoot = KaigiInstructionUtils.canonicalizeOptionalHash(rosterRoot);
      return this;
    }

    public Builder setRosterRoot(final String rosterRootHexOrLiteral) {
      this.rosterRoot = KaigiInstructionUtils.canonicalizeOptionalHash(rosterRootHexOrLiteral);
      return this;
    }

    Builder setRosterRootLiteral(final String literal) {
      this.rosterRoot = literal;
      return this;
    }

    public Builder setProof(final byte[] proofBytes) {
      this.proofBase64 = proofBytes == null ? null : KaigiInstructionUtils.toBase64(proofBytes);
      return this;
    }

    public Builder setProofBase64(final String proofBase64) {
      this.proofBase64 =
          proofBase64 == null ? null : KaigiInstructionUtils.requireBase64(proofBase64, "proof");
      return this;
    }

    public JoinKaigiInstruction build() {
      if (callId == null) {
        throw new IllegalStateException("callId must be provided");
      }
      if (participant == null) {
        throw new IllegalStateException("participant must be provided");
      }
      return new JoinKaigiInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      KaigiInstructionUtils.appendCallId(callId, args, "call");
      args.put("participant", participant);
      if (commitment != null) {
        args.put("commitment.commitment", commitment);
        if (commitmentAliasTag != null) {
          args.put("commitment.alias_tag", commitmentAliasTag);
        }
      }
      if (nullifierDigest != null) {
        args.put("nullifier.digest", nullifierDigest);
        if (nullifierIssuedAtMs != null) {
          args.put("nullifier.issued_at_ms", Long.toUnsignedString(nullifierIssuedAtMs));
        }
      }
      if (rosterRoot != null) {
        args.put("roster_root", rosterRoot);
      }
      if (proofBase64 != null) {
        args.put("proof", proofBase64);
      }
      return args;
    }
  }
}

File: .taira-source-c80bf4d85a97/kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/JoinKaigiInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

/** Typed representation of a `JoinKaigi` instruction. */
class JoinKaigiInstruction(
    @JvmField val callId: KaigiInstructionUtils.CallId,
    @JvmField val participant: String,
    @JvmField val commitment: String? = null,
    @JvmField val commitmentAliasTag: String? = null,
    @JvmField val nullifierDigest: String? = null,
    @JvmField val nullifierIssuedAtMs: Long? = null,
    @JvmField val rosterRoot: String? = null,
    @JvmField val proofBase64: String? = null,
) : InstructionTemplate {

    init {
        require(participant.isNotBlank()) { "participant must not be blank" }
        if (nullifierIssuedAtMs != null) {
            require(nullifierIssuedAtMs >= 0) { "nullifier issuedAtMs must be non-negative" }
        }
        if (proofBase64 != null) {
            KaigiInstructionUtils.requireBase64(proofBase64, "proof")
        }
    }

    override val kind: InstructionKind = InstructionKind.CUSTOM

    override val arguments: Map<String, String> by lazy { canonicalArguments() }

    private fun canonicalArguments(): Map<String, String> {
        val args = linkedMapOf<String, String>()
        args["action"] = "JoinKaigi"
        KaigiInstructionUtils.appendCallId(callId, args, "call")
        args["participant"] = participant
        if (commitment != null) {
            args["commitment.commitment"] = commitment
            if (commitmentAliasTag != null) {
                args["commitment.alias_tag"] = commitmentAliasTag
            }
        }
        if (nullifierDigest != null) {
            args["nullifier.digest"] = nullifierDigest
            if (nullifierIssuedAtMs != null) {
                args["nullifier.issued_at_ms"] = java.lang.Long.toUnsignedString(nullifierIssuedAtMs)
            }
        }
        if (rosterRoot != null) {
            args["roster_root"] = rosterRoot
        }
        if (proofBase64 != null) {
            args["proof"] = proofBase64
        }
        return args
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is JoinKaigiInstruction) return false
        return callId == other.callId
            && participant == other.participant
            && commitment == other.commitment
            && commitmentAliasTag == other.commitmentAliasTag
            && nullifierDigest == other.nullifierDigest
            && nullifierIssuedAtMs == other.nullifierIssuedAtMs
            && rosterRoot == other.rosterRoot
            && proofBase64 == other.proofBase64
    }

    override fun hashCode(): Int {
        var result = callId.hashCode()
        result = 31 * result + participant.hashCode()
        result = 31 * result + (commitment?.hashCode() ?: 0)
        result = 31 * result + (commitmentAliasTag?.hashCode() ?: 0)
        result = 31 * result + (nullifierDigest?.hashCode() ?: 0)
        result = 31 * result + (nullifierIssuedAtMs?.hashCode() ?: 0)
        result = 31 * result + (rosterRoot?.hashCode() ?: 0)
        result = 31 * result + (proofBase64?.hashCode() ?: 0)
        return result
    }

    companion object {
        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): JoinKaigiInstruction {
            val callId = KaigiInstructionUtils.parseCallId(arguments, "call")
            val participant = KaigiInstructionUtils.require(arguments, "participant")

            val commitmentValue = arguments["commitment.commitment"]
            val aliasTag = if (commitmentValue != null) arguments["commitment.alias_tag"] else null

            val nullifier = arguments["nullifier.digest"]
            val nullifierIssuedAt = if (nullifier != null) {
                KaigiInstructionUtils.parseOptionalUnsignedLong(
                    arguments["nullifier.issued_at_ms"],
                    "nullifier.issued_at_ms",
                )
            } else {
                null
            }

            val proof = arguments["proof"]

            return JoinKaigiInstruction(
                callId = callId,
                participant = participant,
                commitment = commitmentValue,
                commitmentAliasTag = aliasTag,
                nullifierDigest = nullifier,
                nullifierIssuedAtMs = nullifierIssuedAt,
                rosterRoot = arguments["roster_root"],
                proofBase64 = proof,
            )
        }
    }
}
